contrib/pyzfs/libzfs/models/vdevtree.py

```diff
@@ -0,0 +1,314 @@
+              "end_time=%s "
+              "bytes_to_examine=%d "
+              "bytes_examined=%d "
+              "bytes_to_process=%d "
+              "bytes_processed=%d "
+              "scan_errors=%d "
+              ">" % (
+                  SCAN_STATE(self.scan_state),
+                  self.start_time,
+                  self.end_time,
+                  self.bytes_to_examine,
+                  self.bytes_examined,
+                  self.bytes_to_process,
+                  self.bytes_processed,
+                  self.scan_errors))
+        return ss
+
+    def __init__(self,
+                 pool_scan_func,
+                 scan_state,
+                 start_time,
+                 end_time,
+                 bytes_to_examine,
+                 bytes_examined,
+                 bytes_to_process,
+                 bytes_processed,
+                 scan_errors,
+                 bytes_examined_per_pass,
+                 scan_pass_start_time,
+                 scrub_pass_pause_time,
+                 total_scrub_pause_time_elapsed,
+                 bytes_issued_per_scan,
+                 total_bytes_issued):
+        self.pool_scan_func = pool_scan_func
+        self.scan_state = scan_state
+        self.start_time = start_time
+        self.end_time = end_time
+        self.bytes_to_examine = bytes_to_examine
+        self.bytes_examined = bytes_examined
+        self.bytes_to_process = bytes_to_process
+        self.bytes_processed = bytes_processed
+        self.scan_errors = scan_errors
+        self.bytes_examined_per_pass = bytes_examined_per_pass
+        self.scan_pass_start_time = scan_pass_start_time
+        self.scrub_pass_pause_time = scrub_pass_pause_time
+        self.total_scrub_pause_time_elapsed = total_scrub_pause_time_elapsed
+        self.bytes_issued_per_scan = bytes_issued_per_scan
+        self.total_bytes_issued = total_bytes_issued
+
+    @staticmethod
+    def construct_from_pool_scan_stats(ps_stats):
+        return PoolScanStat(
+            ps_stats[0],
+            ps_stats[1],
+            ps_stats[2],
+            ps_stats[3],
+            ps_stats[4],
+            ps_stats[5],
+            ps_stats[6],
+            ps_stats[7],
+            ps_stats[8],
+            ps_stats[9],
+            ps_stats[10],
+            ps_stats[11],
+            ps_stats[12],
+            ps_stats[13],
+            ps_stats[14]
+        )
+
+
+class VDevTree(object):
+    """
+    We lose a fair bit of information when translating the vdev_tree information
+    given to us by ZFS, as this class does not have all the properties that are present
+    in the ZFS nvlist; instead it just contains the properties we require (for now).
+    """
+    device_type = ""
+    devices = []
+    spares = []
+    l2_cache = []
+    logs = None
+    parity = 0
+    name = ""
+    path = ""
+    vdev_stats = None
+    pool_scan_stats = None
+
+    def __str__(self):
+        repr = ("<VDevTree "
```

Comment:
Missing `dedup` and `special` type

Reply:
I don't see any `dedup` or `special` field in the Vdev nvlist, so I don't quite follow what you mean by this?